Detectors/EMCAL/reconstruction/src/Clusterizer.cxx

```diff
@@ -0,0 +1,180 @@
+  int nSupMod = std::get<0>(cellIndex);
+
+  auto phiEtaIndex = mEMCALGeometry->GetCellPhiEtaIndexInSModule(nSupMod, std::get<1>(cellIndex), std::get<2>(cellIndex), std::get<3>(cellIndex));
+  row = std::get<0>(phiEtaIndex);
+  column = std::get<1>(phiEtaIndex);
+
+  // Add shifts wrt. supermodule and type of calorimeter
+  // NOTE:
+  // * Rows (phi) are arranged that one space is left empty between supermodules in phi
+  //   This is due to the physical gap that forbids clustering
+  // * For DCAL, there is an additional empty column between two supermodules in eta
+  //   Again, this is to account for the gap in DCAL
+
+  row += nSupMod / 2 * (24 + 1);
+  // In DCAL, leave a gap between two SMs with same phi
+  if (!mEMCALGeometry->IsDCALSM(nSupMod)) // EMCAL
+    column += nSupMod % 2 * 48;
+  else
+    column += nSupMod % 2 * (48 + 1);
+}
+
+///
+/// Return number of found clusters. Start clustering from highest energy cell.
+//____________________________________________________________________________
+void Clusterizer::findClusters(const std::vector<Digit>& digitArray)
+{
+  mFoundClusters.clear();
+  mDigitIndices.clear();
+
+  // Algorithm
+  // - Fill digits in 2D topological map
+  // - Fill struct arrays (energy,x,y)  (to get mapping energy -> (x,y))
+  // - Create 2D bitmap (digit is already clustered or not)
+  // - Sort struct arrays with descending energy
+  //
+  // - Loop over arrays:
+  // --> Check 2D bitmap (don't use digit which are already clustered)
+  // --> Take valid digit with highest energy as seed (they are already sorted)
+  // --> Recursive to neighboughs and create cluster
+  // --> Seed cell and all neighbours belonging to cluster will be put in 2D bitmap
+
+  // Reset digit maps and cell masks
+  // Loop over one array dim, then reset each array
+  for (auto iArr = 0; iArr < NROWS; iArr++) {
+    mCellMask[iArr].fill(0);
+    mDigitMap[iArr].fill(0);
+  }
+
+  // Calibrate digits and fill the maps/arrays
+  int nCells = 0;
+  double ehs = 0.0;
+  for (auto dig : digitArray) {
+    Float_t digitEnergy = dig.getEnergy();
+    Float_t time = dig.getTimeStamp();
+
+    if (digitEnergy < mThresholdCellEnergy || time > mTimeMax || time < mTimeMin)
+      continue;
+    if (!mEMCALGeometry->CheckAbsCellId(dig.getTower()))
+      continue;
+
+    ehs += digitEnergy;
+
+    // Put digit to 2D map
+    int row = 0, column = 0;
+    getTopologicalRowColumn(dig, row, column);
+    mDigitMap[row][column] = &dig; // mDigitMap saves pointers to digits, therefore use addr operator here
+    mSeedList[nCells].energy = digitEnergy;
+    mSeedList[nCells].row = row;
+    mSeedList[nCells].column = column;
+    nCells++;
+  }
+
+  // Sort struct arrays with ascending energy
+  std::sort(mSeedList.begin(), std::next(std::begin(mSeedList), nCells));
+  //std::sort(mSeedList, mSeedList+nCells);
+
+  // Take next valid digit in calorimeter as seed (in descending energy order)
+  for (int i = nCells - 1; i >= 0; i--) {
```

Comment:
``for (int i=nCells;i--;) `` does the same but has one if less

Reply:
Done. I tested that this works although I don't fully understand it: Is i-- implementing an implicit break condition?

Reply:
yes: for the last iteration it will check 1-- > 0, execute it for 0 then exit at 0--.